### DISCUSSION TITLE
add_all is executing 1000 single insert instead of one with 1000 parameters

### DISCUSSION THREAD

**Discussion by Bechma** [sqlalchemy/sqlalchemy#7111] [Usage Questions] [answered]
```python
engine = create_async_engine(db_connstring, pool_pre_ping=True)
session = async_scoped_session(
    sessionmaker(bind=engine, expire_on_commit=False, class_=AsyncSession),
    scopefunc=current_task)

class A(Base):
    __tablename__ = "ATable"

    id = Column("ID", Integer, Identity(), primary_key=True)
    ticket = Column("Ticket", Text)
    entity = Column("Entity", Text)

rows = get_a_lot_of_rows()  # 1000 min, 400k max
await session.execute("""TRUNCATE "ATable" RESTART IDENTITY CASCADE;""")
session.add_all(rows)
await session.commit()
```

Tracing SQLAlchemy with opentelemetry:
![image](https://user-images.githubusercontent.com/19294519/135249452-b518f525-db23-4f01-b626-62d5be85db5c.png)

This is slowing the code comparing to the bulk_save_objects or bulk_insert_mappings from the sync scope_session. Is there any way to instruct the engine or session to always do an executemany?

Thanks in advance

**Comment by CaselIT** [comment, marked as answer]
Hi,

What db and driver are you using?
IIRC it's currently supported only on postgresql when using psycopg2. There is an issue to extend it to other db/drivres: https://github.com/sqlalchemy/sqlalchemy/issues/6047

Also given the number here `rows = get_a_lot_of_rows()  # 1000 min, 400k max` you most likely should avoid instantiating the orm classes, and return a list of dict directly, then you can use `session.execute(sa.insert(A), rows)`

**Reply by zzzeek** [reply]
this topic is covered in depth at https://docs.sqlalchemy.org/en/14/faq/performance.html#i-m-inserting-400-000-rows-with-the-orm-and-it-s-really-slow .


are you providing pre-established primary key values?  the ORM otherwise has to get back the PK of every row for every driver other than psycopg2 right now and cannot use executemany().

the best way to insert the rows is to send dictionaries along with an insert() statement, not using session.add().   bulk_execute() is a more complicated way of getting a similar result.

**Reply by Bechma** [reply]
PostgreSQL and asyncpg is the driver I'm using(from what I know, It's the only one supported for async applications).

should I use then always execute(insert, rows) in order to improve performance? The dot notation of the orm classes is very helpful. Is there any programmatic way to convert that orm classes into a dictionary with the keys as the column names? (the \_\_dict\_\_ method returns the python properties names, but no the column names).

**Reply by CaselIT** [reply]
I think if you have a procedure that creates 400k rows, it would be best to avoid instantiating also the orm objects, since that most likely is quite slow and uses a lot of memory compared to just dealing with dicts

**Reply by Bechma** [reply]
From #6935 I understood that add_all had the same performance improvements than bulk_save_objects, but I was completely wrong, my tests for just 4k objects:

- add_all: 90s (async)
- bulk_save_objects: 13s (sync)
- execute insert: 4s (async dict, no orm classes)

I gotta things to change asap, thanks a lot guys
